add developers guide for substreams-sink-kv (#140)
streamingfast/substreams: docs/developers-guide/substreams-sink-kv.md

---
description: StreamingFast Substreams sink kv
---

# `substreams-sink-kv` introduction

## Purpose

This documentation will assist you in using `substreams-sink-kv` to write data from your existing substreams into a key-value store and serve it back through Connect-Web/GRPC.

## Overview

> * `substreams-sink-kv` works by reading the output of specially-designed substreams module (usually called `kv_out`) that produces data in a protobuf-encoded structure called `sf.substreams.sink.kv.v1.KVOperations`.

Reviewer: I would remove the bullet list here personally, might have been copied over from the other tutorial, but it feels weird that an overview is a list.

> * The data is written to a key-value store (ex: badger, bigtable, TiKV)

Reviewer: ```suggestion
* The data is written to a key-value store. Currently supported KV store are Badger, BigTable and TiKV.
```

* A Connect-Web interface makes the data available directly from the `substreams-sink-kv` process. Alternatively, you can consume the data directly from your key-value store.

## Requirements

> * An existing substreams (including substreams.yaml and rust code) that you want to instrument for `substreams-sink-kv`.

Reviewer: ```suggestion
* An existing substreams (including `substreams.yaml` and Rust code) that you want to instrument for `substreams-sink-kv`.
```

> * (Optional) A key-value store where you want to send your data (a badger local file can be used for development or very-small-scale deployments)

Reviewer: I don't see how this is optional

> * Knowledge about substreams development (start [here](https://substreams.streamingfast.io/getting-started/quickstart))

Reviewer: ```suggestion
* Knowledge about Substreams development (start [here](https://substreams.streamingfast.io/getting-started/quickstart))
```

* Rust installation and compiler

## Installation

> * Install [substreams-sink-kv](https://github.com/streamingfast/substreams-sink-kv) latest release

Reviewer: Should point to release page

> * Install [substreams](https://github.com/streamingfast/substreams) latest release

Reviewer: Should point to documentation on Substreams installation

* Install [grpcurl](https://github.com/fullstorydev/grpcurl) to easily read the data back from the KV store

## Quick test of your setup

> * Before you start instrumenting your substreams for the `kv-sink`, it may be a good idea to make sure that you can run your substreams normally:

Reviewer: Useless bullet list

```bash
substreams run -e mainnet.eth.streamingfast.io:443 substreams.yaml some_mapper_module --start-block 1000000 --stop-block +1
```

> > **Note** To connect to a public StreamingFast substreams endpoint, you will need an authentication token, follow this [guide](https://substreams.streamingfast.io/reference-and-specs/authentication) to obtain one.

Reviewer: Useless, we can simply mention that a working Substreams project is expected in `Instrumenting your substreams`

> ## Instrumenting your substreams

Reviewer: ```suggestion
## Instrumenting your Substreams
```

### Assumptions

The following instructions will assume that you are instrumenting [substreams-eth-block-meta](https://github.com/streamingfast/substreams-eth-block-meta), which contains:

> * a store `store_block_meta_end` defined like [this](https://github.com/streamingfast/substreams-eth-block-meta/blob/v0.4.0/substreams.yaml#L29-L34):

Reviewer: ```suggestion
* A store `store_block_meta_end` defined like [this](https://github.com/streamingfast/substreams-eth-block-meta/blob/v0.4.0/substreams.yaml#L29-L34):
```

> ```yaml

Reviewer: I think alignment is wrong, check rendered version probably in GitHub, going to see if rendering seems off or not

# substreams.yaml
...
  - name: store_block_meta_end
    kind: store
    updatePolicy: set
    valueType: proto:eth.block_meta.v1.BlockMeta
    inputs:
      - source: sf.ethereum.type.v2.Block
```

* a `eth.block_meta.v1.BlockMeta` protobuf structure like [this](https://github.com/streamingfast/substreams-eth-block-meta/blob/v0.4.0/proto/block_meta.proto#L7-L12):
```proto
message BlockMeta {
  uint64 number = 1;
  bytes hash = 2;
  bytes parent_hash = 3;
  google.protobuf.Timestamp timestamp = 4;
}
```

> > **Note** The [substreams-eth-block-meta](https://github.com/streamingfast/substreams-eth-block-meta) is already instrumented for sink-kv, the proposed changes here are a simplified version of what has been implemented. Please adjust the proposed code to your own substreams.

Reviewer: ```suggestion
> **Note** The [substreams-eth-block-meta](https://github.com/streamingfast/substreams-eth-block-meta) is already instrumented for sink-kv, the proposed changes here are a simplified version of what has been implemented. Please adjust the proposed code to your own Substreams.
```

> ### Import the cargo module

Reviewer: ```suggestion
### Import the Cargo module
```

1. Add the `substreams-sink-kv` crate to your `Cargo.toml`:

```toml
# Cargo.toml

[dependencies]
substreams-sink-kv = "0.1.1"
# ...

```

> 2. Add a `kv_out` module  to your `substreams.yaml`:

Reviewer: All bullet list should be `1.`, Markdown will automatically adjust the number when rendering (revisit all bullet list in the doc)

```yaml
# substreams.yaml
...
  - name: kv_out
    kind: map
    inputs:
      - store: store_block_meta_end
        mode: deltas
    output:
      type: proto:sf.substreams.kv.v1.KVOperations
```

> 3. Add a `kv_out` public function to your `src/lib.rs`:

Reviewer: ```suggestion
3. Add `map` module implementation function named `kv_out` to your `src/lib.rs`:
```

    // Here, we could add more operations to the kv_ops
    // ...

    Ok(kv_ops)
}
```

4. Add the `kv::process_deltas` transformation function referenced in the last snippet:
```rs
// src/kv_out.rs

use substreams::proto;
use substreams::store::{self, DeltaProto};
use substreams_sink_kv::pb::kv::KvOperations;

use crate::pb::block_meta::BlockMeta;

pub fn process_deltas(ops: &mut KvOperations, deltas: store::Deltas<DeltaProto<BlockMeta>>) {
    use substreams::pb::substreams::store_delta::Operation;

    for delta in deltas.deltas {
        match delta.operation {
            // KV Operations do not distinguish between Create and Update.
            Operation::Create | Operation::Update => {
                let val = proto::encode(&delta.new_value).unwrap();
                ops.push_new(delta.key, val, delta.ordinal);
            }
            Operation::Delete => ops.push_delete(&delta.key, delta.ordinal),
            x => panic!("unsupported opeation {:?}", x),
        }
    }
}
```

## Test your substreams

1. Compile your changes in your rust code:

```
cargo build --release --target=wasm32-unknown-unknown
```

2. Run with `substreams` command directly:

```bash

> substreams run -e mainnet.eth.streamingfast.io:443 substreams.yaml some_mapper_module --start-block 1000000 --stop-block +1

Reviewer: Probably make sense to use `kv_out` instead of `some_mapper_module`?

2023-01-12T10:08:34.186-0500 INFO (sink-kv) substreams sink stats {"db_flush_rate": "10.500 flush/s (21 total)", "data_msg_rate": "0.000 msg/s (0 total)", "progress_msg_rate": "0.000 msg/s (0 total)", "block_rate": "0.000 blocks/s (0 total)", "flushed_entries": 0, "last_block": "None"}
2023-01-12T10:08:34.186-0500 INFO (sink-kv) substreams sink stats {"progress_msg_rate": "16551.500 msg/s (33103 total)", "block_rate": "10941.500 blocks/s (21883 total)", "last_block": "#291883 (66d03f819dde948b297c8d582889246d7ba11a5b947335497f8716a7b608f78e)"}
```

> **Note** This writes the data to a local folder "./badger_data.db/" in Badger format. You can `rm -rf ./badger_data.db` between your tests to cleanup all existing data.

4. Look at the stored data

You can scan the whole dataset using the 'Scan' command:

```bash
grpcurl --plaintext -d '{"begin": "", "limit":100}' localhost:8000 sf.substreams.sink.kv.v1.Kv/Scan
```

You can look at data by key prefix:

```bash
grpcurl --plaintext   -d '{"prefix": "day:first:201511", "limit":31}' localhost:8000 sf.substreams.sink.kv.v1.Kv/GetByPrefix
```

> 

Reviewer: Seems just to be missing some instructions about the Web code. We should show quickly how someone can run the example project with minimal instructions and where in the code he should look for the important pieces.
